### PR TITLE
use-sequenced-action-runs-to-reduce-stages

### PR DESCRIPTION
**Pull request by tekhaus**
In the original task, the stage2 and stage3 preview blocks reassign the event object so that it can just call event.data.collection_id and get the passed collection id.

This reassignment prevents using event.preview blocks later within that event topic block [that stumped me for a bit], so I changed that up in order to have the second preview block in the stage2 topic appear after the logic to assign all_product_ids.
 
An initial approach I considered was to consolidate the preview blocks in stage2 at the top as a matter of convention, but seeing preview data unrelated to the action that initiated stage2 just seemed odd. Plus, I then had to do something dirty like this - {% assign all_product_ids = in_stock_product_ids | concat: out_of_stock_product_ids | default: event.data.ordered_product_ids %}.

### CODE REVIEW

**Comment by isaacbowen** [comment]
:D All of that makes sense, yeah.

Since we're here, let's do this properly. You're right about moving event.preview blocks up as far as possible; for tasks that respond to several event topics, it feels best to me to do this _immediately after_ we've branched for a specific topic. Further, moving action blocks as far _down_ as possible is _also_ good. Combining both of those priorities, how does this structure look?

https://gist.github.com/isaacbowen/e6259dd38269777f283d59d3f33e4a89

Diff with your code: https://gist.github.com/isaacbowen/e6259dd38269777f283d59d3f33e4a89/revisions

(Not super sure where the right place is to render the very final `collectionUpdate` action; it should end up as close to the end of the task as possible while still being before the `collectionReorderProducts` actions.)

> This reassignment prevents using event.preview blocks later within that event topic block [that stumped me for a bit] [...]

haha, indeed. :D There's nothing sacred about this, though, so stubbing for `event.preview` is fine by me. ;) I do that in the code I linked above.

**Comment by tekhaus** [comment]
> Since we're here, let's do this properly. You're right about moving event.preview blocks up as far as possible; for tasks that respond to several event topics, it feels best to me to do this _immediately after_ we've branched for a specific topic.

Sometimes I forget the basic tenet of quality task building... have the stub validate your task logic, don't use it to force the results you need in later actions if your logic should be doing that for you. Thank you for the direction.

File: docs/move-out-of-stock-products-to-the-end-of-a-collection/README.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # Move out-of-stock products to the end of a collection
 
-This task re-sorts your collections, beginning with the sort order of your choice (alphabetically, best selling first, etc), and then moving all out-of-stock products to the very end of the collection.
+This task re-sorts your collections, beginning with the sort order of your choice (alphabetically, best selling first, etc), and then moving all out-of-stock products to the very end of the collection.
```
Comment (isaacbowen) [inline docs/move-out-of-stock-products-to-the-end-of-a-collection/README.md:3]: extraneous space at the end of this line

File: docs/move-out-of-stock-products-to-the-end-of-a-collection/script.liquid [hunk 2]
```diff
@@ -7,6 +7,22 @@
...
 
 {% if event.topic contains "mechanic/" %}
+  {% if event.preview %}
+    {% capture shop_json %}
+      {
+        "collections": [
+          {
+            "id": {{ options.collection_titles_or_ids__array_required.first | json }},
+            "admin_graphql_api_id": "gid://shopify/Collection/1234567890"
+
```
Comment (isaacbowen) [inline docs/move-out-of-stock-products-to-the-end-of-a-collection/script.liquid:17]: extraneous newline

**Review by isaacbowen — CHANGES_REQUESTED** [review]
I'm happy. :) Two whitespace nits, then let's get this merged.

**Review by isaacbowen — APPROVED** [review]

**Pull request merged**